simple_network_sim/loaders.py

```diff
@@ -115,3 +117,172 @@ def readNodeAttributesJSON(filename):
+    represents a given population, and can return the expected number of
+    interactions (per day) a member of this population will have with some
+    other target.
+    """
+    def __init__(self, ages, interactions):
+        """Initialiser. Entries should be a dict mapping an AgeRange to an
+        expected number of interactions."""
+        self._entries = {}
+        for age, interact in zip(ages, interactions):
+            self._entries[age] = float(interact)
+
+    def __getitem__(self, age):
+        """Get the expected number of interactions (per day) that someone from
+        this MixingRow would have with someone with the given age, or age
+        range.
+        """
+        if isinstance(age, str) or isinstance(age, tuple):
+            return self._entries[AgeRange(age)]
+        for key, value in self._entries.items():
+            if age in key:
+                return value
+        raise Exception(f'Could not find {age} in MixingRow')
+
+    def __str__(self):
+        return "[" + ", ".join(f"{str(key)}: {str(val)}"
+                               for key, val in self._entries.items()) + "]"
+
+
+class MixingMatrix:
+    """Can give the expected number of interactions per day a given person will
+    have with some other person, based on the ages of the two people, or given
+    age-ranges. A MixingMatrix will often be instantiated from a CSV file (see
+    initialiser). It can be used as follows:
+
+        mm = MixingMatrix("sample_input_file/sample_20200327_comix_social_contacts.sampleCSV")
+        print(mm[28][57])  # Prints the expected number of interactions a 28 year old
+                           # would have with a 57 year old in a day
+        print(mm["[30,40)"]["70+"])  # Prints the expected number of interactions someone in the
+                                     # age range [30-40) would have with someone aged 70 or older
+                                     # in any given day.
+        print(mm[(30,40)]["70+"])  # As above
+
+    """
+
+    def __init__(self, infile):
+        """Reads an input file. The input file should be a CSV file, with the
+        first row and column as headers. These contain age ranges in either the
+        format "[a, b)", or the format "a+". The entry in the table in row R
+        and column C then indicates how many expected interactions per day a
+        member of the R column interacts with a member of the C columns.
+        """
+        self._matrix = {}
+        with open(infile, "r") as inp:
+            reader = csv.reader(inp)
+            headers = [AgeRange(text) for text in next(reader)[1:]]
+            if len(headers) != len(set(headers)):
+                raise Exception("Duplicate column header found in mixing matrix")
+            # Check for any overlap in the column headers
+            for one in headers:
+                for two in headers:
+                    if one == two:
```

Comment:
minor thing, we could use enumerate(headers) here and compare the headers position rather than value. This way we complain if there are repeated lines too. Eg.:

```
for i,one in enumerate(headers):
    for j,two in enumerate(headers):
        if i != j:
            _check_overlap(one, two)
```

Reply:
Used enumerate to check for equality. Also removed the lines two above, which checked for duplicate entries since that was now redundant.